zebrad/src/components/miner.rs

```diff
@@ -0,0 +1,527 @@
+        );
+
+        // Tell the next get_block_template() call to wait until the template has changed.
+        parameters.long_poll_id = Some(template.long_poll_id);
+
+        let block = proposal_block_from_template(&template, TimeSource::CurTime)
+            .expect("unexpected invalid block template");
+
+        // If the template has actually changed, send an updated template.
+        template_sender.send_if_modified(|old_block| {
+            if old_block.as_ref().map(|b| *b.header) == Some(*block.header) {
+                return false;
+            }
+            *old_block = Some(Arc::new(block));
+            true
+        });
+
+        // If the blockchain is changing rapidly, limit how often we'll update the template.
+        // But if we're shutting down, do that immediately.
+        if !template_sender.is_closed() && !is_shutting_down() {
+            sleep(Duration::from_secs(
+                GET_BLOCK_TEMPLATE_MEMPOOL_LONG_POLL_INTERVAL,
+            ))
+            .await;
+        }
+    }
+
+    Ok(())
+}
+
+/// Runs a single mining thread that gets blocks from the `template_receiver`, calculates equihash
+/// solutions with nonces based on `solver_id`, and submits valid blocks to Zebra's block validator.
+///
+/// This method is CPU and memory-intensive. It uses 144 MB of RAM and one CPU core while running.
+/// It can run for minutes or hours if the network difficulty is high. Mining uses a thread with
+/// low CPU priority.
+#[instrument(skip(template_receiver, rpc))]
+pub async fn run_mining_solver<Mempool, State, Tip, BlockVerifierRouter, SyncStatus, AddressBook>(
+    solver_id: u8,
+    mut template_receiver: WatchReceiver<Option<Arc<Block>>>,
+    rpc: GetBlockTemplateRpcImpl<Mempool, State, Tip, BlockVerifierRouter, SyncStatus, AddressBook>,
+) -> Result<(), Report>
+where
+    Mempool: Service<
+            mempool::Request,
+            Response = mempool::Response,
+            Error = zebra_node_services::BoxError,
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    Mempool::Future: Send,
+    State: Service<
+            zebra_state::ReadRequest,
+            Response = zebra_state::ReadResponse,
+            Error = zebra_state::BoxError,
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    <State as Service<zebra_state::ReadRequest>>::Future: Send,
+    Tip: ChainTip + Clone + Send + Sync + 'static,
+    BlockVerifierRouter: Service<zebra_consensus::Request, Response = block::Hash, Error = zebra_consensus::BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <BlockVerifierRouter as Service<zebra_consensus::Request>>::Future: Send,
+    SyncStatus: ChainSyncStatus + Clone + Send + Sync + 'static,
+    AddressBook: AddressBookPeers + Clone + Send + Sync + 'static,
+{
+    // Shut down the task when the template sender is dropped, or Zebra shuts down.
+    while template_receiver.has_changed().is_ok() && !is_shutting_down() {
```

Comment:
```suggestion
    // TODO: Add an `is_closed` method and replace the `has_changed().is_ok()` with `!is_closed()`
    while template_receiver.has_changed().is_ok() && !is_shutting_down() {
```

Reply:
If you'd like to do this in another PR, go for it!

Reply:
> Update: Nevermind, there's no `is_closed()` method on `watch::Receiver`

It would make the code more readable to add an `is_closed()` method to our `WatchReceiver`, where we control the API.

Reply:
> It would make the code more readable to add an `is_closed()` method to our `WatchReceiver`, where we control the API.

Yes it will, okay, I'll go for it in another PR :D